Arch becomes BIM

yorikvanhavre opened the pull request:
This involves:

* Merging the [BIM workbench](https://github.com/yorikvanhavre/BIM_Workbench) which consists mostly of GUI commands
* Merging the [NativeIFC](https://github.com/yorikvanhavre/FreeCAD-NativeIFC) which is a standalone module
* Moved importers to a subdir
* Removed GUI commands from the Arch* files and put them together with the BIM commands
* Moved all the make*() functions from the Arch* files and added them to Arch.py
* Renaming folders, workbench name and CMake targets from **Arch** to **BIM**

For the user not much will change (they will have the BIM WB menu and toolbars instead of Arch, which has  basically more tools), for developers one should just know Arch is now BIM.

maxwxyz commented:
will the commands and icons have the arch prefix or will they be converted to BIM like it was done with CAM? Or will there be a mix between arch an bim commands?

yorikvanhavre commented:
up to now i tried to be as less disruptive as possible, in case other addons are using arch stuff, and for translations to not break. i was thinking to do all the rest (get rid of the arch name entirely) later on...

furgo16 commented:
This is excellent, looking forward to this!

- What is the plan with the documentation: will the Arch_* wiki pages stay as they are and the new BIM_* namespace created for the new tools?
- Is there any way contributors can help with the transition to the "one and only" BIM workbench?

yorikvanhavre commented:
for the documentation i propose to keep everything as is for now, since the command names didn't change, and we do thât refined transition at a later moment, with more time to adjust translations.

for now, as soon as this is merged, i'd say the most import-nt would be to update the [BIM workbench page](https://wiki.freecad.org/BIM_Workbench), and make sure all the tools are documented

maxwxyz commented:
are there any other icons that need refinement or changes (BIM, Arch, Draft)?

yorikvanhavre commented:
well yes actually 😇 not right now, but the next point on my list is to have an icon for each type of https://standards.buildingsmart.org/IFC/RELEASE/IFC4_1/FINAL/HTML/schema/ifcproductextension/lexical/ifcbuildingelement.htm there are several that have no icon yet, (plate, pile, etc...)

marcuspollio commented:
Hi Yorik,

Wow, nice ! That's a big PR =D

Congrats for reaching this point ! Hopefully, it will be easier to have in one place only all architecture/construction related stuff, for users and developers as well ;D

In this PR, I would suggest :

- Remove all "exp/experimental" tools and features (walls, join, extend, opening, door, window, arch view and offset). There are too many bugs and inconsistency. They are poorly exposed (only in the Toolbars) and not integrated with the whole workbench. It looks like their IFC support does not match the one of the non-experimental equivalents. Also, @carlopav does not look to be active anymore (maybe he can deny or confirm), making maintenance heavier.

- Also remove the (generic) Box tool. There's better way for that and there's a report of a deprecated function (use "DraftGeomUtils.placement_from_points" instead)

In later steps (can be done by other contributors as well) :

- Make the naming consistent and standardized, in particular the whole floor, level, storey confusion so it is compliant with [IFC](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcBuildingStorey.htm). Most importantly the user-facing stuff, ideally also a refactor of the internal code (if it's breaking compatibility, better now than later).

- Rearrange a bit the Toolbars, menus and Task panels in a more streamlined and less overwhelming way (construction sequence and standards). But this can be improved gradually anyways.

> [...] an icon for each type

I can help with that. I did some [Flat-Style icons](https://github.com/obelisk79/FreeCAD-Flat-Icons) last year with more BIM objects, I can adapt them for the current icon style. Let me know which one you would need first (apart from plate and pile).

![Icons-flat](https://github.com/FreeCAD/FreeCAD/assets/131592747/dd42492a-f697-4ad8-831e-c8b382e85e21)


I will give this PR a test when the building tests are all OK.

carlopav commented:
> * Remove all "exp/experimental" tools and features (walls, join, extend, opening, door, window, arch view and offset). There are too many bugs and inconsistency. They are poorly exposed (only in the Toolbars) and not integrated with the whole workbench. It looks like their IFC support does not match the one of the non-experimental equivalents. Also, @carlopav does not look to be active anymore (maybe he can deny or confirm), making maintenance heavier.

Absolutely agree with @marcuspollio. The experimental tools were a funny experiment, but lacked a general vision of the problem, and I think It's pointless to mantain them.

yorikvanhavre commented:
The experimental tools are already out. They are still in the BIM workbench repo though, and they will stay there as I don't plan to remove that repo ever :) @carlopav you might want to back up that code for yourself, who knows there is more to get done with it some day...

The box tool we could remove indeed, but this PR is mostly about transferring the BIM WB "as is" , I think tool removals should be done later, so they can be easily undone if needed.

Same for the Floor/Level/Storey. Indeed, maybe the best name here is then "Storey"?

About general UI layout, I'd be more than happy with a little help ;) Please propose! But yes indeed that's probably matter for later

About icons, this is the whole list we need. I marked those where we already have an icon:

* [ ] IfcBuildingElementProxy
* [ ] IfcCovering
* [x] IfcBeam
* [x] IfcColumn
* [x] IfcCurtainWall
* [x] IfcDoor
* [ ] IfcMember
* [ ] IfcRailing
* [ ] IfcRamp
* [ ] IfcRampFlight
* [x] IfcWall
* [x] IfcSlab
* [ ] IfcStairFlight
* [x] IfcWindow
* [x] IfcStair
* [x] IfcRoof
* [ ] IfcPile
* [ ] IfcFooting
* [ ] IfcBuildingElementComponent
* [ ] IfcPlate

furgo16 commented:
If native IFC is going to become an integral part of BIM in FreeCAD, is it worth shipping IfcOpenShell with the FreeCAD builds as well, to make life easier for users?

marcuspollio commented:
> The experimental tools are already out.

Sorry, my bad ! I haven't compiled it yet =D

Sure, this PR is big enough, all these further improvements can come later.

> I'd be more than happy with a little help ;)

I've asked @maxwxyz if he can create a new BIM Label et Project so we can add issues to that (such as Tools, Icons, UI, etc.) Do you mind ?

yorikvanhavre commented:
> I've asked @maxwxyz if he can create a new BIM Label et Project so we can add issues to that (such as Tools, Icons, UI, etc.) Do you mind ?

Of course not! Glad with any help :)

maxwxyz commented:
We can use the arch label. I'll rename it after the merge. I cannot setup a project, this has to be done by an owner.

yorikvanhavre commented:
There is still a problem with this PR... After compiling it, the Arch_rc.py resource file fails to load with a `QResource: error decompressing zlib content (-3)` error when starting FreeCAD, and the BIM icon subsequently does not appear in the Start page, and the BIM WB fails to load and FreeCAD crashes.

Could anyone test this too?

The funny thing is, compiling Arch_rc.py manually with `rcc -g python --compress-algo=zlib --compress=1 -o Arch_rc.py src/Mod/BIM/Resources/Arch.qrc` produces a valid file that works. And rcc is the same in both cases...

furgo16 commented:
> Could anyone test this too?

I can confirm this behavior. I've wanted to test this branch for days, but I've not been able to because of this error (the BIM workbench would not load). I've not looked at it in detail, though.

The error I got was related to Arch_rc, but I don't have a current build to show the exact error. I'll start one to check.

yorikvanhavre commented:
The same error is what's provoking the failed CI build on ubuntu 22.04.
There seems to be a problem somewhere with some Qt version. I'll try to find out what

furgo16 commented:
If that helps, I just built the branch again and here's the output. This time the error is slightly different for me:

```
During initialization the error "UI file does not exist" occurred in $HOME/dev/freecad-source/build/Mod/Arch/InitGui.py
Please look into the log file for further information
During initialization the error ""'BIMWorkbench' already exists."" occurred in $HOME/.local/share/FreeCAD/Mod/BIM/./InitGui.py
Please look into the log file for further information
Cannot find icon: $HOME/dev/freecad-source/build/share/Mod/BIM/Resources/icons/BIMWorkbench.svg
Cannot find icon: $HOME/dev/freecad-source/build/share/Mod/Web/Resources/icons/WebWorkbench.svg
```

1. FreeCAD starts and shows the above message
2. The BIM workbench is nowhere to be found (neither in settings nor in the TabBar)

I'm not sure if this is a red herring and unrelated to the `QResource: error decompressing zlib content (-3)` error, though. I have the `freecad-daily` PPA installed on Ubuntu 22.04, and I'm wondering if there is a clash with the external BIM workbench files installed from there and the now built-in BIM workbench files from this local build.

Note that I'm not running any install target. I simply start the binary from ` $HOME/dev/freecad-source/build/bin`.

furgo16 commented:
I'm also getting this error when switching workbenches:

```python-traceback
Cannot find icon: BIM_Welcome.svg
14:40:53  Cannot find icon: BIM_Help
14:40:53  Cannot find icon: BIM_Tutorial
14:40:53  Cannot find icon: BIM_Help
14:44:19  Traceback (most recent call last):
14:44:19    File "$HOME/dev/freecad-source/build/Mod/Arch/nativeifc/ifc_observer.py", line 240, in propose_conversion
14:44:19      dlg = FreeCADGui.PySideUic.loadUi(
14:44:19    File "<string>", line 4, in <module>
14:44:19  RuntimeError: Cannot open file
14:44:19  Traceback (most recent call last):
14:44:19    File "$HOME//dev/freecad-source/build/Mod/Arch/nativeifc/ifc_observer.py", line 240, in propose_conversion
14:44:19      dlg = FreeCADGui.PySideUic.loadUi(
14:44:19    File "<string>", line 4, in <module>
14:44:19  RuntimeError: Cannot open file
```

furgo16 commented:
Yes, it's there: `$INSTALL_PREFIX/usr/local/Mod/BIM/ARch_rc.py`

Unrelated, but worth noting: it's the first time I've ran the `make install` target. Until now, I just executed the built binary from the `build/bin` folder. I now installed it with `make DESTDIR=$HOME/opt install`. That seemed to work, but I cannot launch the binary (`error while loading shared libraries: libFreeCADGUI.so: cannot open shared object file: No such file or directory`), so I cannot really test the BIM workbench changes from that location.

yorikvanhavre commented:
That's strange... I can run FreeCAD from both the build and install dirs here.

In any case we do have a problem with the resource file. It's unloadable for some reason, and since it contains all the icons and ui files, that's why you are getting those "file not found" errors.

Still don't understand why when I run the exact same commands as the build process does, it produces a valid resource file...

furgo16 commented:
I can run the installed binary now. I simply used the default install target with no parameters, which installs FreeCAD on `/usr/local`. I guess it didn't like to be relocated after running configure. 

So, back to the original issue, I can also reproduce it. As soon as I try to switch to the BIM workbench, I'm welcomed with these messages. The workbench then fails to load:

![image](https://github.com/FreeCAD/FreeCAD/assets/148809153/643d4ecd-e67a-4a4e-b24d-9f5a0ec55908)


```python-traceback
21:02:33  No module named 'Arch_rc'
21:02:33  Traceback (most recent call last):
  File "$HOME/.local/share/FreeCAD/Mod/BIM/./InitGui.py", line 77, in Initialize
    import DraftTools
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/usr/local/Mod/Draft/DraftTools.py", line 49, in <module>
    import Draft
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/usr/local/Mod/Draft/Draft.py", line 53, in <module>
    from draftutils.utils import ARROW_TYPES as arrowtypes
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/usr/local/Mod/Draft/draftutils/utils.py", line 42, in <module>
    from draftutils import params
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/usr/local/Mod/Draft/draftutils/params.py", line 31, in <module>
    import Arch_rc
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
```

yorikvanhavre commented:
Found it! The culprit is this line: https://github.com/FreeCAD/FreeCAD/blob/main/cMake/FreeCAD_Helpers/SetupShibokenAndPyside.cmake#L265

Apparently by chance our new Arch_rc.py file contains a line that is caught and suppressed by that sed command...

Now I understand this line was there to remove a header printed by PyRCC that contained a date. The new Qt5+ RCC does print this:

```python
# Resource object code (Python 3)
# Created by: object code
# Created by: The Resource Compiler for Qt version 5.15.10
# WARNING! All changes made in this file will be lost!
```

That has no more date. I think we could just comment out that line...

furgo16 commented:
Great! After testing, I'm getting two errors (running from the build folder):

1. After executing `build/bin/FreeCAD`:
   ```
   During initialization the error ""'BIMWorkbench' already exists."" occurred in 
   $HOME/dev/FreeCAD/build/Mod/Arch/InitGui.py
   Please look into the log file for further information
   ```
   I'm not sure if this is related to having an installation of BIM as an external workbench from the freecad-daily PPA, and the built-in and external BIMs now collide? I'll try to remove the external BIM next.
   Also the error message mentions a log file, but it doesn't mention where or which log file that is:

2. When the GUI loads, switching to the BIM workbench:
   ```python-traceback
   module 'ArchWindow' has no attribute '_CommandWindow'
   Traceback (most recent call last):
     File "/home/dpm/.local/share/FreeCAD/Mod/BIM/./InitGui.py", line 80, in Initialize
       import BimCommands
     File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
       return original_import(name, *args, **kwargs)
     File "/home/dpm/.local/share/FreeCAD/Mod/BIM/./BimCommands.py", line 315, in <module>
       class BIM_Door(ArchWindow._CommandWindow):
   ```

furgo16 commented:
Removing the external BIM workbench from fixed both errors from the previous comment. Now I'm getting different errors:

1. When executing FreeCAD:
   ```
   During initialization the error "UI file does not exist" occurred in $HOME/dev/FreeCAD/build/Mod/Arch/InitGui.py
   Please look into the log file for further information
   ```

2. When switching to the BIM workbench:
   ```
   Cannot find icon: $HOME/dev/FreeCAD/build/share/Mod/BIM/Resources/icons/BIMWorkbench.svg
   Cannot find icon: Sketch
   Cannot find icon: Arch_Project_IFC
   Cannot find icon: BIM_Column
   Cannot find icon: BIM_Beam
   Cannot find icon: BIM_Slab
   Cannot find icon: BIM_Door
   Cannot find icon: BIM_Box
   Cannot find icon: BIM_Library
   Cannot find icon: BIM_DimensionAligned
   Cannot find icon: BIM_DimensionHorizontal.svg
   Cannot find icon: BIM_DimensionVertical
   Cannot find icon: BIM_Leader
   Cannot find icon: techdraw-PageDefault
   Cannot find icon: techdraw-ArchView
   Cannot find icon: BIM_Copy
   Cannot find icon: BIM_Clone
   Cannot find icon: BIM_Unclone
   Cannot find icon: BIM_Rewire
   Cannot find icon: BIM_Glue
   Cannot find icon: BIM_Reextrude
   Cannot find icon: Tree_Part
   Cannot find icon: BIM_Views
   Cannot find icon: BIM_Project
   Cannot find icon: BIM_Windows
   Cannot find icon: BIM_IfcElements
   Cannot find icon: BIM_IfcQuantities
   Cannot find icon: BIM_IfcProperties
   Cannot find icon: BIM_Classification
   Cannot find icon: BIM_Layers
   Cannot find icon: BIM_Material
   Cannot find icon: BIM_Preflight
   Cannot find icon: BIM_TogglePanels
   Cannot find icon: BIM_Trash
   Cannot find icon: BIM_WPView
   Cannot find icon: BIM_Diff
   Cannot find icon: BIM_Welcome.svg
   Cannot find icon: BIM_Help
   Cannot find icon: BIM_Tutorial
   Cannot find icon: BIM_Help
   ```
   
3. When creating a new document and zooming out:
   ```python-traceback
   Traceback (most recent call last):
     File "$HOME/dev/FreeCAD/build/Mod/Arch/nativeifc/ifc_observer.py", line 240, in propose_conversion
       dlg = FreeCADGui.PySideUic.loadUi(
     File "<string>", line 4, in <module>
   RuntimeError: Cannot open file
   ```

I used:

```
OS: Ubuntu 22.04.4 LTS (ubuntu:GNOME/ubuntu)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.37213 +10 (Git)
Build type: Unknown
Branch: pr-13783-new
Hash: f3c619086ae2180bcc3f3dd765ffe8b1ecb31ba8
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 9.1.0, OCC 7.6.3
Installed mods: 
  * OpenTheme 2024.5.3
  * parts_library
```

furgo16 commented:
Errors 1. and 2. from the previous comment go away when running the make install target and executing FreeCAD from `/usr/local/bin/FreeCAD`.

Error 3. remains.

furgo16 commented:
Some quick testing:

1. I noticed the IFC Project command has been replaced by the native IFC project command, so I tested it:
   ![image](https://github.com/FreeCAD/FreeCAD/assets/148809153/2a1b474b-dff9-4eec-8814-7c46c51eb68f)
   I'm not sure if it's intended, but everything under this new native IFC project container loses its icon, which is replaced by a generic IFC icon
2. I was having some difficulties simply adding a window to a wall. It just didn't work as usual. I had to do a manual addition to the wall, and that did not work either. When I created the window, I got the `  <class 'NameError'>` error on the report view. No further context.
[test.zip](https://github.com/FreeCAD/FreeCAD/files/15234734/test.zip)

yorikvanhavre commented:
To summarize the problems above:

1) Indeed if the BIM addon is installed, it conflicts with this one. I now [added code](https://github.com/yorikvanhavre/BIM_Workbench/commit/02b509d6a66d5ad62a8d5b903dc5219dd7c27ff0) to the BIM addon, if it detects the new BIM workbench, you'll get a message telling you to uninstall the BIM (and NativeIFC) addon.

2) If you are building FreeCAD over an existing build, you will still have an Arch workbench module from previous installs. This indeed creates conflicts too, as the Arch_rc resource file is picked from the wrong folder. So all the "file not found" errors are likely coming from there. You basically need to remove the Mod/Arch folder in your install folder. This is only a problem if you build FreeCAD yourself, though, so not sure we need to do something about it...

3) Indeed the Arch project tool is replaced by the IFC project tool. Basically because there is no real need to use a Project if it's not for using IFC. This was the case in the "old" BIM WB already

4) Now is where the fun begins :sweat_smile: There are certainly MANY bugs left. I will welcome all the help possible to thoroughly test everything :) But I'd wait until this is merged and open separate issues, to not loose track.

furgo16 commented:
> If you are building FreeCAD over an existing build, you will still have an Arch workbench module from previous installs. This indeed creates conflicts too, as the Arch_rc resource file is picked from the wrong folder. So all the "file not found" errors are likely coming from there. You basically need to remove the Mod/Arch folder in your install folder. This is only a problem if you build FreeCAD yourself, though, so not sure we need to do something about it...

```
Traceback (most recent call last):
  File "/usr/local/Mod/BIM/nativeifc/ifc_observer.py", line 240, in propose_conversion
    dlg = FreeCADGui.PySideUic.loadUi(
  File "<string>", line 4, in <module>
RuntimeError: Cannot open file
```

This was on a clean install (I had not done any installation from source on my desktop machine previously). I'll keep an eye on it. As per the comment below, happy to report it as a separate issue later if it persists.

> Now is where the fun begins 😅 There are certainly MANY bugs left. I will welcome all the help possible to thoroughly test everything :) But I'd wait until this is merged and open separate issues, to not loose track.

Happy to see this merged soon, and report issues afterwards. Thanks!

Other than that, and unless specific testing is required, at a quick glance, it works :-)

marcuspollio commented:
Hi @yorikvanhavre
Thank you for solving the build issue, it compiles on my side now =D

When testing, I cannot create an IFC Project # Gui.runCommand('BIM_Project',0) as there's an error with IfcOpenShell.
When using the former BIM external workbench (addon), I worked out of the box.
Should I install it or tweak something ?

<details>
<summary>Report</summary>

```
Cannot find icon: /home/***/share/Mod/BIM/Resources/icons/BIMWorkbench.svg
Wizard shaft module cannot be loaded
Traceback (most recent call last):
File "/home/***/Mod/Arch/nativeifc/ifc_observer.py", line 240, in propose_conversion
    dlg = FreeCADGui.PySideUic.loadUi(
  File "<string>", line 4, in <module>
RuntimeError: Cannot open file
Running the Python command 'BIM_Project' failed:
Traceback (most recent call last):
  File "/home/***/Mod/Arch/bimcommands/BimProject.py", line 45, in Activated
    from nativeifc import ifc_tools
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/home/***/Mod/Arch/nativeifc/ifc_tools.py", line 32, in <module>
    from importers import exportIFC
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/home/***/Mod/Arch/importers/exportIFC.py", line 42, in <module>
    from importers import exportIFCHelper
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/home/***/Mod/Arch/importers/exportIFCHelper.py", line 27, in <module>
    import ifcopenshell
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)

No module named 'ifcopenshell'

```
</details>

<details>
<summary>Full version info</summary>

```shell
OS: GNU/Linux Ubuntu 22.04 LTS based (DE:GNOME/pop)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.37213 +11 (Git)
Build type: Release
Branch: pr/13783
Hash: 98844db62dd45d53abdfddfaec2df04b199de637
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk , OCC 7.5.1
```
</details>

Thank you ! ;D

yorikvanhavre commented:
> `File "/home/***/Mod/Arch/nativeifc/ifc_observer.py", line 240, in propose_conversion`

There is something very weird here... It should be Mod/BIM/nativeifc... not Mod/Arch/nativeifc... Looks like you have a mixed or partial build.

I would suggest removing entirely the Mod/Arch folder, then rebuild

marcuspollio commented:
Hi @yorikvanhavre 

Ok, so solved the issue with IfcOpenShell.
I've redone a complete new build (removed all previous files, also config, local, cache, etc.), but still have bizarre stuff with this Arch-BIM mix-up ¯\_(ツ)_/¯

> There is something very weird here... It should be Mod/BIM/nativeifc... not Mod/Arch/nativeifc

Indeed, it would make sense. But see below :

<details><summary> Build folder </summary>

![build](https://github.com/FreeCAD/FreeCAD/assets/131592747/430799c2-c270-4214-8978-1f05154732df)

</details>

<details><summary> Report </summary>

```

Cannot find icon: /PATH/build-bim/share/Mod/BIM/Resources/icons/BIMWorkbench.svg

Wizard shaft module cannot be loaded

Traceback (most recent call last):
  File "/PATH/build-bim/Mod/Arch/nativeifc/ifc_observer.py", line 240, in propose_conversion
    dlg = FreeCADGui.PySideUic.loadUi(
  File "<string>", line 4, in <module>
RuntimeError: Cannot open file
Running the Python command 'BIM_ProjectManager' failed:
Traceback (most recent call last):
  File "/PATH/build-bim/Mod/Arch/bimcommands/BimProjectManager.py", line 60, in Activated
    self.form.buildingUse.addItems(Arch.BuildingTypes)
module 'Arch' has no attribute 'BuildingTypes'

Traceback (most recent call last):
  File "/PATH/build-bim/Mod/Arch/bimcommands/BimProject.py", line 45, in Activated
    from nativeifc import ifc_tools
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/PATH/build-bim/Mod/Arch/nativeifc/ifc_tools.py", line 41, in <module>
    from ifcopenshell.util import placement

  File "/PATH/.local/lib/python3.10/site-packages/ifcopenshell/util/placement.py", line 23, in <module>
    MatrixType = np.ndarray[np.ndarray[float]]

'type' object is not subscriptable

```

</details>


<details><summary> Cmake settings </summary>

```

@:~/PATH/build-bim$ cmake  -DCMAKE_BUILD_TYPE=Release -DBUILD_FEM=OFF -DBUILD_CAM=OFF -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -G Ninja ../FreeCAD
-- Compiler: GNU, version: 11.4.0
-- prefix: /usr/local
-- bindir: bin
-- datadir: share
-- docdir: share/doc/FreeCAD
-- includedir: include
-- libdir: lib
-- cmake: 3.22.1
-- Compiling with Qt 5
-- BUILD_MESH_PART requires BUILD_SMESH to be ON, but it is ""
-- BUILD_FLAT_MESH requires BUILD_MESH_PART to be ON, but it is "OFF"
-- BUILD_OPENSCAD requires BUILD_MESH_PART to be ON, but it is "OFF"
-- Checking for connection to GitHub...
-- GitHub connection established for FetchContent
-- Module support is disabled.
-- Version: 9.1.0
-- Build type: Release
-- CXX_STANDARD: 17
-- Required features: cxx_variadic_templates
-- fmt was downloaded using FetchContent into /PATH/build-bim/_deps/fmt-src
-- Found Python3: /usr/bin/python3.10 (found version "3.10.12") found components: Interpreter Development Development.Module Development.Embed 
-- PyCXX found:
--   Headers:  /PATH/FreeCAD/src
--   Sources:  /PATH/FreeCAD/src/CXX
--   Version:  7.1.9
-- -- Found OCE/OpenCASCADE version: 7.5.1
-- -- OCE/OpenCASCADE include directory: /usr/include/opencascade
-- -- OCE/OpenCASCADE shared libraries directory: /usr/lib/x86_64-linux-gnu
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found suitable version "1.74.0", minimum required is "1.65") found components: filesystem program_options regex system thread date_time 
-- Set up to compile with Qt 5.15.3
-- Shiboken2Config: Using default python: .cpython-310-x86_64-linux-gnu
-- SHIBOKEN_PYTHON_INCLUDE_DIRS computed to value: '/usr/include/python3.10'
-- SHIBOKEN_PYTHON_LIBRARIES computed to value: ''
-- libshiboken built for RelWithDebInfo
-- PYTHON_CONFIG_SUFFIX: .cpython-310-x86_64-linux-gnu
-- libshiboken built for RelWithDebInfo
-- PySide 5.15.2 Python module found at /usr/lib/python3/dist-packages/PySide2.

-- Found Matplotlib: /usr/lib/python3/dist-packages/matplotlib (found version "3.5.1")
-- Platform is 64-bit, set -D_OCC64
-- [E57] Revison ID: E57Format-2.2.1-x86_64-linux-gcc11
-- [E57] Building static library
-- [OndselSolver] Building shared library
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.74.0/BoostConfig.cmake (found version "1.74.0") found components: python310 
-- found Boost: 1_74
-- boost-incude dirs are: /usr/include
-- boost-python lib is: 
-- boost_LIBRARY_DIRS is: /usr/lib/x86_64-linux-gnu
-- Boost_LIBRARIES is: Boost::python
-- area module (for Path Workbench) will be installed to: lib
-- Performing Test _flag_found
-- Performing Test _flag_found - Success
-- Performing Test _flag_found
-- Performing Test _flag_found - Success
-- Coin3D doc is not installed
-- Found Python3: /usr/bin/python3.10 (found version "3.10.12") found components: Interpreter 

   ==============
     System
   ==============

-- bindir:                      bin
-- BLAS:                        -undefined-
-- CMAKE_VERSION:               3.22.1
-- Compiler:                    /usr/bin/c++ (11.4.0)
-- datadir:                     share
-- docdir:                      share/doc/FreeCAD
-- includedir:                  include
-- libdir:                      lib
-- prefix:                      /usr/local
-- Python:                      3.10.12 [/usr/bin/python3.10] Suffix: [.cpython-310-x86_64-linux-gnu]

   ==============
     Config
   ==============

-- BUILD_ADDONMGR:              ON
-- BUILD_ASSEMBLY:              ON
-- BUILD_BIM:                   ON
-- BUILD_CLOUD:                 OFF 
-- BUILD_DRAFT:                 ON
-- BUILD_DRAWING:               OFF 
-- BUILD_FEM:                   OFF 
-- BUILD_GUI:                   ON
-- BUILD_HELP:                  ON
-- BUILD_IDF:                   ON
-- BUILD_IMPORT:                ON
-- BUILD_INSPECTION:            ON
-- BUILD_JTREADER:              OFF 
-- BUILD_MATERIAL:              ON
-- BUILD_MEASURE:               ON
-- BUILD_MESH:                  ON
-- BUILD_MESH_PART:             OFF 
-- BUILD_OPENSCAD:              OFF 
-- BUILD_PART:                  ON
-- BUILD_PART_DESIGN:           ON
-- BUILD_PATH:                  ON
-- BUILD_PLOT:                  ON
-- BUILD_POINTS:                ON
-- BUILD_REVERSEENGINEERING:    ON
-- BUILD_ROBOT:                 ON
-- BUILD_SANDBOX:               OFF 
-- BUILD_SHOW:                  ON
-- BUILD_SKETCHER:              ON
-- BUILD_SMESH:                 -undefined-
-- BUILD_SPREADSHEET:           ON
-- BUILD_START:                 ON
-- BUILD_SURFACE:               ON
-- BUILD_TECHDRAW:              ON
-- BUILD_TEST:                  ON
-- BUILD_TUX:                   ON
-- BUILD_VR:                    OFF 
-- BUILD_WEB:                   ON
-- CMAKE_BUILD_TYPE:            Release
-- CMAKE_CXX_FLAGS:             -Wall -Wextra -Wno-write-strings 
-- CMAKE_CXX_STANDARD:          17
-- CMAKE_INSTALL_PREFIX:        /usr/local
-- CMAKE_PREFIX_PATH:           -undefined-
-- ENABLE_DEVELOPER_TESTS:      ON
-- FREECAD_CREATE_MAC_APP:      -undefined-
-- FREECAD_LIBPACK_USE:         OFF 
-- FREECAD_QT_VERSION:          Auto
-- FREECAD_USE_EXTERNAL_KDL:    OFF 
-- FREECAD_USE_EXTERNAL_SMESH:  OFF 
-- FREECAD_USE_FREETYPE:        ON
-- FREECAD_USE_PYBIND11:        OFF 
-- PYTHON_EXECUTABLE:           /usr/bin/python3.10
-- PYTHON_LIBRARY:              /usr/lib/x86_64-linux-gnu/libpython3.10.so
-- USE_CUDA:                    -undefined-
-- USE_OPENCV:                  -undefined-

   ==============
     Libraries
   ==============

-- Boost:                       1.74.0
-- Coin3D:                      4.0.0 [/usr/lib/x86_64-linux-gnu/libCoin.so] [/usr/include]
-- Coin3D_DOC:                  not found
-- DesignerPlugin:              not built (BUILD_DESIGNER_PLUGIN is OFF)
-- Doxygen:                     1.9.1 Language: English
-- Eigen3:                      3.4.0
-- fmt:                         Sources downloaded to /PATH/build-bim/_deps/fmt-src
-- Freetype:                    2.11.1
-- Matplotlib:                  3.5.1 PathDirs: /usr/lib/python3/dist-packages/matplotlib
-- NETGEN:                      not enabled
-- OCC:                         7.5.1 [/usr/lib/x86_64-linux-gnu] [/usr/include/opencascade]
-- OCC_Libs:                    [TKFillet;TKMesh;TKernel;TKG2d;TKG3d;TKMath;TKShHealing;TKXSBase;TKBool;TKBO;TKBRep;TKTopAlgo;TKGeomAlgo;TKGeomBase;TKOffset;TKPrim;TKHLR;TKFeat;TKIGES;TKSTL;TKSTEPBase;TKSTEPAttr;TKSTEP209;TKSTEP]
-- OpenGL_Lib:                  [/usr/lib/x86_64-linux-gnu/libGL.so]
-- OpenGLU_Incl:                [/usr/include]
-- OpenGLU_Lib:                 [/usr/lib/x86_64-linux-gnu/libGLU.so]
-- pivy:                        0.6.5
-- pybind11:                    not enabled
-- PYCXX:                       7.1.9 Incl: /PATH/FreeCAD/src Src:/PATH/FreeCAD/src/CXX
-- PySide:                      5.15.2 [/usr/lib/../include/PySide2]
-- PySideTools:                 v: 2  uic: [/usr/lib/qt5/bin/uic]  rcc: [/usr/lib/qt5/bin/rcc]
-- QtConcurrent:                5.15.3
-- QtCore:                      5.15.3
-- QtNetwork:                   5.15.3
-- QtOpenGL:                    5.15.3
-- QtPrintSupport:              5.15.3
-- QtSvg:                       5.15.3
-- QtTest:                      5.15.3
-- QtUiTools:                   5.15.3
-- QtWidgets:                   5.15.3
-- QtXml:                       5.15.3
-- Rift:                        not enabled (BUILD_VR)
-- Shiboken:                    5.15.2 [/usr/lib/../include/shiboken2;/usr/include/python3.10]
-- SMESH:                       do not build
-- SPNAV:                       [/usr/lib/libspnav.so] [/usr/include]
-- SWIG:                        4.0.2
-- XercesC:                     3.2.3 [/usr/lib/x86_64-linux-gnu/libxerces-c.so] [/usr/include]
-- yaml-cpp:                    0.7.0
-- ZLIB:                        1.2.11

=================================================
Now run 'cmake --build /PATH/build-bim' to build FreeCAD
=================================================

-- Configuring done
-- Generating done
-- Build files have been written to: /PATH/build-bim

```

</details>

Sorry to bother you with this, I'd just like to test properly =D

furgo16 commented:
In my case,

1. I can only run the BIM workbench successfully when doing a `make install` and executing `/usr/local/bin/Freecad`. When running from the build folder via `build/bin/FreeCAD`, it does not find the resources. Not an issue, just a note. And just confirming what I had observed after a few builds/installs.
2. Trying to do what @marcuspollio did results in an error as well. That is, a) start FreeCAD, b) switch to the BIM workbench c) create an empty NativeIFC BIM project via `BIM_PRoject`

```python-traceback
18:33:41  Running the Python command 'BIM_Project' failed:
Traceback (most recent call last):
  File "/usr/local/Mod/BIM/bimcommands/BimProject.py", line 46, in Activated
    project = ifc_tools.create_document_object(FreeCAD.ActiveDocument)
  File "/usr/local/Mod/BIM/nativeifc/ifc_tools.py", line 90, in create_document_object
    ifcfile, project, full = setup_project(obj, filename, shapemode, silent)
  File "/usr/local/Mod/BIM/nativeifc/ifc_tools.py", line 146, in setup_project
    if not "IfcFilePath" in proj.PropertiesList:

'NoneType' object has no attribute 'PropertiesList'
```

I'm not sure if that's the intended workflow, or if an empty FreeCAD (.FCStd) project needs to be created before clicking on `BIM_Project`

```
pip list | grep ifcopenshell
ifcopenshell             0.7.0.240423
```

```
OS: Ubuntu 22.04.4 LTS (ubuntu:GNOME/ubuntu)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.28856
Build type: Unknown
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 9.1.0, OCC 7.6.3
Installed mods: 
  * OpenTheme 2024.5.3
  * parts_library
```

yorikvanhavre commented:
> Indeed, it would make sense. But see below :
> Build folder 

I think the Arch folder in your build folder is a leftover from a previous build, no? In any case, if the Arch folder is there, it is loaded BEFORE the BIM folder, and therefore the Arch_rc file is the wrong one and the resources are not found. We could of course rename Arch_rc to BIM_rc to avoid that but it would then mean more code to change in the original Arch code, and I found safer to change as little as possible there in this PR so everything keeps rebaseable, etc.

But I'm now rebuilding from scratch to test, maybe there is something easier that we can do

> Trying to do what @marcuspollio did results in an error as well. That is, a) start FreeCAD, b) switch to the BIM workbench c) create an empty NativeIFC BIM project via BIM_PRoject

I can't reproduce that error... Mind to show me a screenshot of your nativeIFC prefs screen under menu Edit -> Preferences -> Import/export -> NativeIFC ?

yorikvanhavre commented:
... and the next one should fix the project tool (provisorily - still needs a good refactor that unifies the SingleDoc pref parameter and the IFC icon on the status bar)

furgo16 commented:
I can confirm that in this branch FreeCAD runs off the build directory now :tada: 

I saw the update in your comment, and that you can reproduce the `PropertiesList` issue. Since I was already doing a build, here's the screencapture of my NativeIFC preferences, if that helps. As far as I know, they're the defaults. I did not change them, at least knowingly.

![image](https://github.com/FreeCAD/FreeCAD/assets/148809153/28cbdca6-f44f-4a46-a571-d488383c1fe6)

furgo16 commented:
Now I'm getting this error when clicking on the Project command on an empty document:

```python-traceback
16:22:27  Traceback (most recent call last):
16:22:27    File "$HOME/dev/freecad-source/build/Mod/BIM/nativeifc/ifc_observer.py", line 240, in propose_conversion
16:22:27      dlg = FreeCADGui.PySideUic.loadUi(
16:22:27    File "<string>", line 4, in <module>
16:22:27  RuntimeError: Cannot open file
16:29:36  Running the Python command 'BIM_Project' failed:
Traceback (most recent call last):
  File "$HOME/dev/freecad-source/build/Mod/BIM/bimcommands/BimProject.py", line 45, in Activated
    from nativeifc import ifc_tools
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "$HOME/dev/freecad-source/build/Mod/BIM/nativeifc/ifc_tools.py", line 41, in <module>
    from ifcopenshell.util import placement
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "$HOME/.local/lib/python3.10/site-packages/ifcopenshell/util/placement.py", line 23, in <module>
    MatrixType = np.ndarray[np.ndarray[float]]

'type' object is not subscriptable
```

```
OS: Ubuntu 22.04.4 LTS (ubuntu:GNOME/ubuntu)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.37261 +13 (Git)
Build type: Unknown
Branch: pr-13783
Hash: fcd94202343208001e09e06257b9f3b6f44ee40f
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 9.1.0, OCC 7.6.3
```

furgo16 commented:
> Your IfcOpenShell version is tool old. But I'm working on that too :)

I'm using the latest IfcOpenShell from PyPI, is the BIM workbench expecting a pre-release version?

```
 pip list | grep ifcopenshell
ifcopenshell           0.7.0.240423
```

A few other things I noticed, using an empty document:

1. Pressing the **BIM_IfcProperties** toolbar button:
```python-traceback
17:31:36  Running the Python command 'BIM_IfcProperties' failed:
Traceback (most recent call last):
  File "$HOME/dev/freecad-source/build/Mod/BIM/bimcommands/BimIfcProperties.py", line 143, in Activated
    self.form.tree.selectionModel.selectionChanged.connect(self.updateProperties)

'builtin_function_or_method' object has no attribute 'selectionChanged'
```

2. Pressing the **BIM_Classification** toolbar button:
```python-traceback
Running the Python command 'BIM_Classification' failed:
Traceback (most recent call last):
  File "$HOME/dev/freecad-source/build/Mod/BIM/bimcommands/BimClassification.py", line 55, in Activated
    import BimMaterial
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)

No module named 'BimMaterial'
```

3. Pressing the **BIM_Layers** toolbar button:
```python-traceback
Running the Python command 'BIM_Layers' failed:
Traceback (most recent call last):
  File "/home/dpm/dev/freecad-source/build/Mod/BIM/bimcommands/BimLayers.py", line 86, in Activated
    self.dialog.buttonAssign.setIcon(QtGui.QIcon(":/icons/button_right.svg"))

'PySide2.QtWidgets.QDialog' object has no attribute 'buttonAssign'
```

yorikvanhavre commented:
@furgo16 

> File "$HOME/.local/lib/python3.10/site-packages/ifcopenshell/util/placement.py"

This could indicate you are using not the pip version but a home compiled version... Could you try this?

`import ifcopenshell; print ifcopenshell.version`

For the rest:

> 1. Pressing the BIM_IfcProperties toolbar button

Fixed in above commit

> 2. Pressing the BIM_Classification toolbar button

Fixed in above commit

> 3. Pressing the BIM_Layers toolbar button

I can't reproduce this... I don't understand what's the problem here, there is well a buttonAssign in that dialog

> 4. Pressing the BIM_Material toolbar button, then OK, then removing the created material from the tree view

That's the same problem as above, tool old ifcopenshell version (ifcopenshell/util/placement.py too old)

furgo16 commented:
> This could indicate you are using not the pip version but a home compiled version... Could you try this?

> `import ifcopenshell; print(ifcopenshell.version); print(ifcopenshell.__file__)`

@yorikvanhavre, sure:

```python
>>> import ifcopenshell
>>> print(ifcopenshell.version)
v0.7.0-f7c03db75
>>> print(ifcopenshell.__file__)
$HOME/.local/lib/python3.10/site-packages/ifcopenshell/__init__.py
```

I did not (knowingly at least!) compile ifcopenshell myself. I installed it with pip (as a user, not system-wide, if that makes a difference). The `.local/lib` location is `pip`'s default installation location for user installs, IIRC. The lines below seem to confirm this:

```shell
$ pip show -f ifcopenshell

Name: ifcopenshell
Version: 0.7.0.240423
Summary: Python bindings, utility functions, and high-level API for IfcOpenShell
Home-page: 
Author: 
Author-email: Dion Moult <dion@thinkmoult.com>
License: 
Location: $HOME/.local/lib/python3.10/site-packages
```

One thing I noticed was that the `import ifcopenshell` statement on FreeCAD's console took quite a while to return (10-15 seconds). Doing the same outside FreeCAD with `ipython3` did not have this delay.

yorikvanhavre commented:
That's really weird... Here with the same version as yours, and no crash:
```
>>> import ifcopenshell
>>> ifcopenshell.version
'v0.7.0-f7c03db75'
>>> from nativeifc import ifc_tools
>>> from ifcopenshell.util import placement
```
I also have the 10-seconds delay, though. I suppose the pip version is not optimally compiled...

furgo16 commented:
I can reproduce the crash on FreeCAD's Python console. I'm running from the build directory, having pulled the latest branch commits:

```python-traceback
>>> import ifcopenshell
>>> print(ifcopenshell.version)
v0.7.0-f7c03db75
>>> print(ifcopenshell.__file__)
$HOME/.local/lib/python3.10/site-packages/ifcopenshell/__init__.py
>>> from nativeifc import ifc_tools
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/home/dpm/dev/freecad-source/build/Mod/BIM/nativeifc/ifc_tools.py", line 41, in <module>
    from ifcopenshell.util import placement
  File "/usr/lib/python3/dist-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/home/dpm/.local/lib/python3.10/site-packages/ifcopenshell/util/placement.py", line 23, in <module>
    MatrixType = np.ndarray[np.ndarray[float]]
TypeError: 'type' object is not subscriptable
```

yorikvanhavre commented:
Hm the error actually comes from numpy (File "$HOME/.local/lib/python3.10/site-packages/ifcopenshell/util/placement.py", line 23)

Could you try this?

```python
>>> import numpy
>>> numpy.version.full_version
'1.26.4'
>>> numpy.__file__
'/usr/lib/python3/dist-packages/numpy/__init__.py'
```

Note that indeed the pip version of IfcOpenShell is outdated (mine that I build does not have that numpy line anymore). That's annoying... Means it's likely to be problematic to recommend the pip version

furgo16 commented:
I have indeed an older numpy version than yours:

```
>>> import numpy
>>> numpy.version.full_version
'1.21.5'
>>> numpy.__file__
'/usr/lib/python3/dist-packages/numpy/__init__.py'
```

Let me see if I can find out why it's like that on my system

furgo16 commented:
>> Pressing the BIM_Layers toolbar button

> I can't reproduce this... I don't understand what's the problem here, there is well a buttonAssign in that dialog

Regarding the other issues, I can confirm that the fixes worked. Only the one about BIM_Layers still remains.

furgo16 commented:
Having upgraded `numpy`, I can no longer reproduce the crash. I noticed I had pinned it in the past, to [work around an issue with `OpenCV`](https://forum.freecad.org/viewtopic.php?p=727181&sid=f68d5db3b76e6c5c0b3fb51ff7491b8a#p727181). The majority of users won't have done it, so it won't be an issue.

Regarding the outdated version of `ifcopenshell` on PyPI, do the maintainers there need to be given a nudge? They seem to release the packages fairly often (last one last month), but I cannot see how the pip releases correlate with the upstream releases
- https://pypi.org/project/ifcopenshell/#history
- https://github.com/IfcOpenShell/IfcOpenShell/releases?q=ifcopenshell&expanded=true

What's weird is that the last release in GitHub is from Nov '22, whereas the pip package is Apr '24. Are the pip packagers releasing git snapshots? Have the ifcopenshell people given up on releases?

yorikvanhavre commented:
Probably the pip version is built on the nov'22 code (last release)... That would explain it. In any case, yes, we can suppose the pip version should be working.

furgo16 commented:
The IfcOpenShell docs state that the pip packages are built monthly (I guess from git?) => https://docs.ifcopenshell.org/ifcopenshell-python/installation.html#pypi

At least there was a build last month. But there are some gaps: not all months have a pip release: https://pypi.org/project/ifcopenshell/#history, nor is the release date consistent. So I'm not sure if those monthly builds are automatic or manual.

Alternatively, can the FreeCAD packages ship a known-well-working version of IfcOpenShell?

yorikvanhavre commented:
They do, but I think it's also the pip version...

paullee0 commented:
I use AppImage on Fedora Linux for a long time, it has IfcOpenShell 'built-in'.  Hope this arrangement would not changed :D

paullee0 commented:
Looks like there is a very good progress here, congrats!


Can 'Arch' be somehow be kept in the workbench name?  Arch-BIM? 

Arch is a much much more commonly understood by the general public.  'BIM' is more cryptic, maybe used for another 10 years, the 'auto' company may spin new terms for more business opportunity.  My two cents :)

yorikvanhavre commented:
I was going to call you here to have a look, @paullee0  :) (And @Roy-043 )

Yes the idea is to not change anything to how IfcOpenShell is detected/used: The one you have on your system, or the one that comes with your FreeCAD package. But additionally, there is now a tool that can install it from pip too.

About BIM, there is also a bit of a political stance there: The term BIM is ours, it's not a propriety of big vendors. BIM is at the end nothing else than CAD for the AEC sector (architecture, engineering and construction), we should use it and claim it for ouselves :) Also because it's more inclusive than "Arch", it's not only for architecture. And also, same as "Path" that was renamed to CAM, even if path was more understandable, it's important we use the correct technical terms when we have the possibility. What we need to make clear to the user, and that's the important point, is: BIM is for you, it's not dark magic for skilled wizards. So we must work on making those tools easy to discover and use

yorikvanhavre commented:
I think all issues are currently solved @furgo16 , aside from the BIM_Layers one, but I propose to merge already, to not wait with this any longer, then we will quickly see if others have the same problem, and find a fix easier.

paullee0 commented:
> I was going to call you here to have a look, @paullee0 :) (And @Roy-043 )
> 
> Yes the idea is to not change anything to how IfcOpenShell is detected/used: The one you have on your system, or the one that comes with your FreeCAD package. But additionally, there is now a tool that can install it from pip too.
> 
> About BIM, there is also a bit of a political stance there: The term BIM is ours, it's not a propriety of big vendors. BIM is at the end nothing else than CAD for the AEC sector (architecture, engineering and construction), we should use it and claim it for ouselves :) Also because it's more inclusive than "Arch", it's not only for architecture. And also, same as "Path" that was renamed to CAM, even if path was more understandable, it's important we use the correct technical terms when we have the possibility. What we need to make clear to the user, and that's the important point, is: BIM is for you, it's not dark magic for skilled wizards. So we must work on making those tools easy to discover and use

Thanks for the invitation:)  And glad to know IfcOpenShell should hopefully shipped with AppImage :dagger: 

And about 'BIM', it just missed the dimension of meaning it associated with art, diluted the role of Architect with capital 'A' LOL.  No intention to keep on arguing about the history and theory of Architecture, understood the community is okay with it :D

Would help to test after this is merged, thanks :D

furgo16 commented:
> I think all issues are currently solved @furgo16 , aside from the BIM_Layers one, but I propose to merge already, to not wait with this any longer, then we will quickly see if others have the same problem, and find a fix easier.

@yorikvanhavre absolutely, ship it! :-)

marcuspollio commented:
Hi @yorikvanhavre 

Finally solved my strange issues, no more "Mod/Arch" folders.
But have the same errors as @furgo16 above (numpy, shiboken) ...

Otherwise, just wondering about those tools (missing from BIM WB) :

Structural System : should we use the Axis property of e.g. Beam and Column instead ?
Equipment tools : should we use Component instead ?

yorikvanhavre commented:
@marcuspollio I have a new PR ready right after this one is merged, that reformats a bit the toolbars... I propose we merge this one and then we discuss the toolbar layouts, what goes in, what stays out, okay? Indeed there are some Arch tools missing from BIM. I'd be in favour of reintegrating everything for 1.0 (all Arch tools + all BIM tools) first, to help people transition, then in  layer step discuss if tools should be removed

furgo16 commented:
This is unrelated to the merging of this PR, but since the conversation started, and to keep everything in one place, here is some context on IfcOpenShell versions:

## Summary

| Package name | Forge | Package version | Upstream version | Notes |
| --- | --- | ---| --- | --- |
| IfcOpenShell | [PyPI](https://pypi.org/project/ifcopenshell/) | `0.7.0.240423` | `v0.7.0-f7c03db75` | Package creation date: 2024-04-23. Corresponds to this [upstream commit](https://github.com/IfcOpenShell/IfcOpenShell/commit/f7c03db75). Not shipped with FreeCAD, at least not with .deb and .snap Linux packages |
| IfcOpenShell | [Conda](https://anaconda.org/conda-forge/ifcopenshell) | `v0.7.0.231127` | `0.7.0` | Package creation date: 2023-11-27. As shipped with FreeCAD's weekly AppImage |

If I understand it correctly, the AppImage bundles the `IfcOpenShell` package from Conda Forge. What puzzles me is that the pip version seems to be newer than the conda version, but I think @yorikvanhavre you mentioned that in your system it's the other way around (i.e. pip's version is older)?

Interestingly, IfcOpenShell upstream have their own dedicated registry/archive/forge at https://anaconda.org/ifcopenshell/ifcopenshell - there you can download IfcOpenShell 0.8.0alpha1

## Version details

### IfcOpenShell from PyPI (as of today)
```python
>>> import ifcopenshell
>>> print(ifcopenshell.version)
v0.7.0-f7c03db75
>>> print(ifcopenshell.__file__)
$HOME/.local/lib/python3.10/site-packages/ifcopenshell/__init__.py
```
```
$ pip show -f ifcopenshell

Name: ifcopenshell
Version: 0.7.0.240423
```

### IfcOpenShell from AppImage (FreeCAD_weekly-builds-37213-conda-Linux-x86_64-py311.AppImage)
```python
>>> import ifcopenshell
>>> print(ifcopenshell.version)
0.7.0
>>> print(ifcopenshell.__file__)
/tmp/.mount_FreeCAtqrYCd/usr/lib/python3.11/site-packages/ifcopenshell/__init__.py
```

yorikvanhavre commented:
The problem is that ifcopenshell.verision does not reflect the same thing in all cases... Maybe @Moult or @aothms could help us here? What version should we rely on? Conda or pip? Or any other?

yorikvanhavre commented:
Thanks @Moult !

yorikvanhavre commented:
Okay it's merged!

paullee0 commented:
Some more findings - Regression?

It seems some recent commit/PR merged get reverted for some reason?

e.g.
ArchWall.py, geometry.py

- https://github.com/FreeCAD/FreeCAD/pull/12603
- https://github.com/FreeCAD/FreeCAD/pull/13648

Thanks :)

Roy-043 commented:
There are some rough edges, as would be expected with such a big PR.

While looking at the Arch_CutPlane issue reported on the [forum](https://forum.freecad.org/viewtopic.php?t=87750), I noticed that BimCutPlane.py has a 2024 copyright notice that does not mention the authors of the original ArchCutPlane.py file. This seems strange. The code of BimCutPlane.py is clearly copied from the old ArchCutPlane.py file. I assume this is a copy-paste error.

Maybe this needs checking for all files?

yorikvanhavre commented:
> A dialog box comes up and request user to remove original external BIM, Native-IFC add-on.

This is indeed annoying and not well handled. I'm thinking the Tux WB could do that work to detect the "wrong" BIM WB and warn the user. Uninstalling the BIM (And NativeIFC) addon is very simple using the addon manager. If the user installed BIM or NativeIFC manually, without the addon manager, I guess they will know how to remove it as well...

> It seems the External Add-on is not working to remove them, ordinary users might not know where to delete the external add-on folders

It should, though... I'll give it a test

> The button to initiate 'Strict IFC' not working, it calls the IfcImport and then IfcExport dialog box, and the cursor keep show the rotating / waiting icons, needs to kill 'x' the dialog box to continue

This whole thing is indeed not well implemented yet ( see #14123 ). I'll remove that functionality for now, and reimplement better afterwards

> It seems some recent commit/PR merged get reverted for some reason?

The first one is there: https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/BIM/ArchWall.py#L633
The second one indeed I don't see.. Weird, I had rebased everything just before merging the PR. Maybe they crossed each other. Anyway, I've reapplied it with #14186

> While looking at the Arch_CutPlane issue reported on the [forum](https://forum.freecad.org/viewtopic.php?t=87750), I noticed that BimCutPlane.py has a 2024 copyright notice that does not mention the authors of the original ArchCutPlane.py file. This seems strange. The code of BimCutPlane.py is clearly copied from the old ArchCutPlane.py file. I assume this is a copy-paste error.

Indeed... What happened is that the Arch* files stayed in place, for backwards compatibility, so old files still open without problems. Only the command definition classes were removed from them and placed in new files in bimcommands/ The ArchCutPlane however, only contained a command and no object, do the original file is gone, and I didn't take enough care with the header... I'll fix that now - #14187

I believe it's the only one where that happened, fortunately.

paullee0 commented:
Noted, not sure in this big merge or afterwards :
- ArchBuilding.py no longer present for some reasons

yorikvanhavre commented:
> ArchBuilding.py no longer present for some reasons

Correct. It was not used anymore, as the Building is now simply a BuildingPart with its IFC type set as "IfcBuilding". I would like to do the same for sites too.

furgo16 commented:
> Correct. It was not used anymore, as the Building is now simply a BuildingPart with its IFC type set as "IfcBuilding"

How will this work for existing models that used the current ArchBuilding and sites? Will they be automatically upgraded to  BuildingPart with IFC type set as _IfcBuilding_, etc?

yorikvanhavre commented:
They already use BuildingParts... the old building object is obsoleted since a long time, before 0.20 I think

paullee0 commented:
BIM and Native-IFC entries are still in Add-on Manager, cause some confusion.

Should they be removed from Add-on Manager list now, or it needs for previous user to uninstall?

yorikvanhavre commented:
No because people using 0.21 still need them. I think they should stay for some time after the release, but they need a big warning in their README. But I think that should be done at release time

Note that the NativeIFC is not in the addon manager, unless you added it manually

yorikvanhavre commented:
Installation problems are discussed further on https://github.com/FreeCAD/FreeCAD/issues/14072
